### DISCUSSION TITLE
Any best practice to create a demo project?

### DISCUSSION THREAD

**Discussion by huxuan** [copier-org/copier#1314] [Q&A]
I wonder whether if there is any best practice for demonstrating a copier template project. The most important features would be auto update alongside the template. And considering there may be different `choices` and the template configuration might also change from time to time, it might be better to manage the `answers` in somewhere for different settings.

**Comment by yajo** [comment]
I guess you could record a video or asciinema tutorial showing how your template behaves.

**Reply by huxuan** [reply]
Actually I want to demonstrate the template rendering results, especially for several kinds of typical settings. It can be different repos or different branches in the same repo. And there might be a regular GitHub Actions / GitLab CI/CD to update the repo. Any ideas or suggestions?

**Reply by pawamoy** [reply]
You can probably get inspiration from existing templates?

I'll list a few here:

- my own Python project template: https://github.com/pawamoy/copier-pdm
- PDM's own project template: https://github.com/pdm-project/copier-pdm

**Reply by huxuan** [reply]
Thanks for the reference. I really appreciate those existing template and really learned a lot from them. But what I need might be an actual working example, because the template I created has many CI/CD related features which supports both GitHub Actions and GitLab CI/CD, such as lint, test, documentation and docs/pages. Moreover, there are some recommended project/group settings alongside the template.

**Reply by pawamoy** [reply]
If I understand correctly, you're looking for best-practices for *how to demo Copier features*. And you say you already have created a template. At work once, I had to demo Copier and a template I had created. It was more about the features the template provided rather than Copier itself. But to demo Copier features, I would have done this:

1. generate a new project from template, tag N, answering questions
1. show the generated project structure a bit, with answers injected in files
1. modify some files in the projects, files that are initially templated, and commit
1. update the project from template, tag N+1, skipping questions
1. show that project is updated, while previous modifications from step 3 are kept
1. update project again, this time giving a few different answers
1. show updated project

---

If that's not what you meant, then I'm confused :sweat_smile: Maybe you meant, *how to create a minimal template that nicely demonstrates Copier features*? In that case, any minimal template will do: single templated file, default answers file :shrug:

**Reply by huxuan** [reply]
Hi @pawamoy, thanks for your informative reply and sorry for the potential misleading.

I think what I want to demonstrate is the features provided by the **template** rather than copier. Or in other words, I want to propose a kind of best practice for team projects, including various kinds of integrations, configurations and common used modules.

You can find the repo here, though it is still in an early stage: https://github.com/serious-scaffold/serious-scaffold-python

**Reply by pawamoy** [reply]
I see, thanks for clarifying!

Well, you could start with a high-level view of your template and the workflow it offers (something like this: https://pawamoy.github.io/copier-pdm/work/#workflow). Then you can sequentially unfold everything, following the workflow, explaining at each step why this is considered best practice (benefits of formatting code with black, of typing with mypy, of linting with ruff, of testing and reporting coverage with pytest+coverage, of using GH actions, etc.).

**Reply by huxuan** [reply]
Hi @pawamoy, thanks for your suggestions. I will working on the documentation later. For the working example, I will also give it a try, will come back for further suggestions then.

**Comment by coretl** [comment]
I've tried creating a working example with https://github.com/DiamondLightSource/python-copier-template. On a push to main, then it renders the template with some default values, and uses an SSH deploy key to push to https://github.com/DiamondLightSource/python-copier-template-example. This seems to work nicely, apart from not reflecting an example repo CI failure back up to the template CI

**Comment by yajo** [comment]
Check out https://github.com/KyleKing/copier-template-tester
